feat(new source): RFC for an Opentelemetry traces source (#11802)
vectordotdev/vector: rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

    mode: grpc

transforms:
  set_key:
    type: remap
      source: |
        if exists!(.tags.user_id) {
          return
        }
        key = get_enrichment_table_record!("api_keys", { "user": .tags.user_id })
        set_dd_api_key(key) # this does not exists yet
      inputs:
        - otlp_traces

sinks:
  dd_trace:
    type: datadog_traces
    default_api_key: 12345678abcdef
    inputs:
      - set_key
```

This demux/conditional action can be seen as an extension of what currently exists in Vector. Other kind of conditional
action like the `filter` transform to discard traces base on certain metadata can be considered to be very similar, as
this also involve evaluating a VRL condition on traces. The key problem here is to exposes traces and spans field in a
way that the user can still manipulate those easily.

This however raises the case of the granularity of a single event ; for instance multiple traces can bundles into a
single payload in both OpenTelemetry and Datadog wire format. Enabling clear processing withou ambiguity advocate for a
clear constraint that should be enforced by all future traces sources : **a single Vector event shall not hold data

> relative to more that one trace**.

Reviewer: It's not clear to me how we will handle this?

If a payload contains multiple traces does the source just split this up into multiple events? Is there any requirement that the traces are together in the same payload when sent out from the sink? If so, that could get complicated, if not then we can probably relax!

Author: AFAIK there is no requirements send traces that would have been ingested together in the same batch from a sink perspective. I'll make that explicit that there is no reason to keep traces together.

- Use the Opentelemetry trace format as the common denominator and base the Vector internal representation to ensure :
  - A clear reference point for conversion between trace formats
  - Avoid destructive manipulation by transforms and keep traces object fully functionnal even after heavy modifications
    while flowing throw the topology

### Traces normalization/format enforcement

For cross format operation like `opentelemetry_traces` source to `datadog_traces` sinks or the opposite (Datadog to
OpenTelemetry) trace standardization is require so between sinks/sources traces will follow one single universal
representation, there is two major possible approach:

  1. Stick to a `LogEvent` based representation and leverage [Vector event schema][schema-work]
  2. Move traces away from their current representation (as LogEvent) and build a new container based on a set of
     dedicated structs representing traces and spans with common properties and generic key/value store(s) to allow a
     certain degree of flexibility.

The second option would have to provide a way to store, at least, all fields from both Opentelemetry and Datadog Traces:

> Datadog [newer trace format][otlp-trace-proto-def] (condensed):

Author: It makes sense, I will relocate those subsections.

    All the expected behaviours regarding the kind of requests/responses code and sequence are clearly defined as well
    as the default URL path (`/v1/traces` for traces).
- Internal traces representation/normalization, two options are opened see [outstanding
  questions](#outstanding-questions), but the consensus is leaning towards a new dedicated container that would:
  - Move away from from the current implementation that relies on `LogEvent` to a dedicated container, the
    implementation will stay  in [./lib/vector-core/src/event/trace.rs][current-trace-in-vector]
  - Borrow most of its semantic from the Opentelemetry Traces format, the `TraceEvent` would then evolve toward a
    concrete dedicated container and contains a slice of spans, the span being directly inspired by the Opentelemtry
    specification.
  - Some top-level information (Like trace ID, trace-wide tags/metrics, the original format)
  - Trace would not get native `VrlTarget` representation anymore, there is a bigger discussion there that should
    probably be adressed separately. As an interim measure few fields may be exposed (like trace ID & trace-wide

>     tags).

Author: My point was that I'm not exactly sure what should be exposed right now, the way I see it would be to at least allow vrl conditions to be evaluated (lets say on "well known" fields like the trace_id or tags/metadata) for filtering/routing and defer the big picture (as in traces vs. remap transform) to a subsequent RFC.

>     tags).

Reviewer 2: Yeah, I can see that.

- APM stats computation:
  - Implement a similar logic that the one done in the Datadog OTLP exporter, this would allow user to use multiple
    Datadog product with Opentelemetry traces and get the same consistent behaviour in all circumstances. APM stats
    computation is hooked [there][apm-stats-computation] in the Datadog exporter. But as this is go code it relies on
    the [Agent codebase][agent-code-for-otlp-exporter] to do the [actual computation][agent-handle-span].
  - Where the APM stats computations is still under discusion, see [outstanding questions](#outstanding-questions)

## Rationale

- Opentelemetry is the de-facto standard for traces, so supporting it at some point is mandatory. Note that this
  consideration is wider than just traces as metrics (and logs) are addressed by the Opentelemetry project.

## Drawbacks

> N/A

Author: Most of the traces format designed for remote submission & processing seems to be able to fit into an opentelemetry inspired container. There are some trace receiver alreadry explicitly doing that [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver) for AWS X-ray and Skywalking traces. Other trace formats that target remote processing like the [new relic format](https://docs.newrelic.com/docs/distributed-tracing/trace-api/report-new-relic-format-traces-trace-api/) seems to be suitable to fit inside an opentelemetry based container. However if we think about trace format like perf data or CTF, that are mostly processed locally, they are really different, and if we want to support that kind of traces this will probably required dedicated container, and this is indeed a drawback, I will update the RFC to reflect that.

## Prior Art

N/A

## Alternatives

- We could keep the Datadog trace-agent as an OTLP->Datadog traces converter and ingest datadog traces from there
- We could keep the Datadog exporter as an OTLP->Datadog traces converter and ingest datadog traces from there
- We could write a Vector exporter for the Opentelemetry collector, note that this would likely leverage the Vector
  protocol and this logic could be applied to metrics as well

## Outstanding Questions

- Do we want to have a single `opentelemetry` source with names output or multiple sources for Opentelemetry metrics and

>   Opentelemetry logs, this is more a user experience discussion as this doesn't have big implementation impact.

Author: The collector has the "processor" construct, which is similar in concept to vector transform (somehow) but there are only few ones available (https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/README.md). On wire the Opentelemetry ecosystem keeps metrics/traces/logs on different gRPC services / HTTP request paths (https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto/collector). That being said having a single source listening for multiple gRPC services or demuxing request based on path seems also point towards a single `opentelemetry` that would emit metrics/traces/logs on relevant named outputs. I will advocated for that in the RFC.

>   Opentelemetry logs, this is more a user experience discussion as this doesn't have big implementation impact.

Reviewer 3: I like the parallel with `datadog_agent` 👍

- APM stats computation:
  - Either in all traces sources (to be done for each source, except for the `datadog_agent` sources where APM stats may

>     be decoded from received payloads) - likely to be the preferred solution

Author: I agree, those metrics can be useful outside datadog, I think this can be generic enough especially when we account for the trace normalisation effort. Also I think this stats computation should come with an opt in/out flag to enable/disable it upon user needs.

  - Either in a transform like `traces_to_metrics`
  - Or in the `datadog_traces` sources

> - The extend of transforms support and VRL (at least condition evaluation)

Reviewer 3: I _think_ this is probably answered?

  - Or with the `trace-agent` [configured to receive OpenTelemetry traces][otlp-traces-with-dd-agent] (both gRPC and HTTP
    transport layer are supported)

### Usecases

As the whole traces processing inside Vector is pretty new, documenting confirmed and most credible use cases in the
near future will help to ensure changes will be implemented so they will be really useful to potential users. This also
help to build something flexible enough to accommodate future needs.

One identified scenario is to demux a trace flow based on some conditions that could be evaluated against any metadata
for a single trace, a group of traces or per spans. From a config perspective this would expect to be functional with
the following configuration:

```yaml
[...]
sources:
  otlp:
    type: opentelemetry
    address: "[::]:8081"
    mode: grpc
    traces_stats: true # Traces stats would be compliant with APM stats

transforms:
  set_key:
    type: remap
      source: |
        if exists!(.tags.user_id) {
          return
        }
        key = get_enrichment_table_record!("api_keys", { "user": .tags.user_id })

>         set_dd_api_key(key) # this does not exists yet

Reviewer 4: This does exist as `set_metadata_field`.

      inputs:
       - otlp.traces # Would exclusively emit traces
       - otlp.traces_stats # Would exclusively emit metrics

sinks:
  dd_trace:
    type: datadog_traces
    default_api_key: 12345678abcdef
    inputs:
      - set_key
```

This demux/conditional action can be seen as an extension of what currently exists in Vector. Other kind of conditional
action like the `filter` transform to discard traces base on certain metadata can be considered to be very similar, as
this also involve evaluating a VRL condition on traces. The key problem here is to exposes traces and spans field in a
way that the user can still manipulate those easily.

> This however raises the case of the granularity of a single event ; for instance multiple traces can bundles into a

Reviewer 3: ```suggestion
This however raises the case of the granularity of a single event ; for instance multiple traces can bundle into a
```

  out-of-the-box with minimal config.

## Proposal

### User Experience

- User would point OpenTelemetry tracing lib directly to a local Vector deployement
- Vector would be configured with a minimal config looking like:

```yaml
sources:
  otlp:
    type: opentelemetry
    address: "[::]:8081"
    mode: grpc
    traces_stats: true # Traces stats would be compliant with APM stats

sinks:
  dd_trace:
    type: datadog_traces
    default_api_key: 12345678abcdef
    inputs:
     - otlp.traces # Would exclusively emit traces
     - otlp.traces_stats # Would exclusively emit metrics
```

And it should just work.

### Implementation

Based on the [usecases](#usecases) previously detailed the implementation will we can extract the following
top-level requirements:

- A Vector trace event shall only contain data relative to one single trace, i.e. traces sources shall create one event
  for each indivual trace ID and its associated spans and metadata.
- Use the Opentelemetry trace format as the common denominator and base the Vector internal representation to ensure :
  - A clear reference point for conversion between trace formats

>   - Avoid destructive manipulation by transforms and keep traces object fully functionnal even after heavy modifications
>     while flowing throw the topology

Reviewer 3: ```suggestion
  - Avoid destructive manipulation by transforms and keep traces object fully functional even after heavy modifications
    while flowing through the topology
```

#### Source structure

> A new `opentelemetry` source sources with a named ouptut `traces` (future extension would cover `metrics` then `logs`):

Reviewer 3: ```suggestion
A new `opentelemetry` source with a named output `traces` (future extension would cover `metrics` then `logs`):
```

- The gRPC variant would use Tonic to spawn a gRPC server (like the `vector` source in its v2 variation) and directly

>   use the [offical gRPC service definitions][otlp-grpc-def], only the traces gRPC service will be accept, this should be

Reviewer 3: ```suggestion
  use the [offical gRPC service definitions][otlp-grpc-def], only the traces gRPC service will be accepted, this should be
```

JSON) and include the resulting value into the tags (a.k.a Meta) map.

This makes the opposite conversion a bit complicated if we want it to be completely symetrical but there was already an
[attempt][otlp-dd-trace-receiver] allow Datadog traces ingestion in the OpenTelemetry collector. While this PR was
closed unmerged this provide a valuable example. Anyways the [otlp-and-other-formats][OpenTelemetry] acknowledges that
some of the OpenTelemetry contruct ends up being stored as tags or annotations in other formats.

Anyway the OpenTelemtry to Datadog traces conversion is dictacted by existing implementation in both the `trace-agent`
and the Datadog exporter as users will expect a consistent behaviour from one solution to another. The same
consideration applies for APM stats computation, as [official implementations][apm-stats-computation] already provides a
reference that define what should be done to get the same result with Vector in the loop. The other way, from Datadog to
OpenTelemetry is less common as of today but while implementing conversions we should ensure that the following path is

> at best idempotent but at least fully functional:

Reviewer 4: Do you have any specific concerns about the parts that won't be idempotent?

Author: I double checked both data structure, I don't find anything that would prevent the double conversions to be idempotent. I will expand a bit on that to clearly highlight that there is not blocking point and that the implementation will enforce that.

Reviewer 4: Great!

supported on both sides working on better interoperability on that particular common ground would likely be a better
option.

**Conclusion**: the implementation will stay around [./lib/vector-core/src/event/trace.rs][current-trace-in-vector], it
will borrow most of the OpenTelemetry to allow straightforward trace conversion to the newer Vector internal
representation. Regarding `datadog_agent` source and `datadog_traces` sink the conversion to/from this newer trace representation will follow existing logic and ensure that standard usecases (like introducing Vector between the Datadog intake and the `trace agen`) do not signigicantly change the end-to-end behaviour. Some top-level information (Like trace ID, trace-wide tags/metrics, the original format) are likely to be added to the internal trace representation for efficiency and convenience.
Trace would not get native `VrlTarget` representation anymore, there is a bigger discussion there that should probably
be adressed separately. As an interim measure few fields may be exposed (At least trace ID & trace-wide tags), the spans
list will not be exposed initially.

> #### APM stats computation

Reviewer 4: Have you considered the alternative of, instead of having the sources do these calculations of APM, have them done in the `datadog_traces` sink? I'm wondering:

* If the metrics are useful outside of the context of Datadog. A naive reading of them makes it seem like this in a Datadog specific enhancement.
* If users would do transformations that would affect these metrics. For example, by filtering out traces. If a user filters out traces, would we still want the APM metrics to reflect all of them?

Reviewer 4: After thinking about this some more. I am actually wondering if this isn't better as a separate transform to give users more control over when this calculation happens. For example, maybe they'd like to filter before calculating the stats? Maybe I'm investing non-existent use cases though 😄 

What if we started with a transform and then added it to the trace sources later? They could just integrate it by running the transform directly in the source.

Author: This is IMHO a relevant use case (maybe not immediately). If you imagine a trace transform that would limit the depth of a trace by removing span deeper than a certain threshold, computing APM stats after this transform is a good idea as it would prevent APM stats to be computed for dropped spans. Also in case of heavy tag manipulation it's probably safe to assume that doing that once and then compute APM stats is probably better than doing the same operation on both the trace flow and the APM stats flow.

Implementation-wise this is probably a good idea to keep that in mind and make it flexible enough (in the end this will always be calculating APM stats on traces using the internal vector representation) so we should be able to implement both with the same underlying implementation and we may ultimately keep both option: e.g. a pass-through configuration would enabled APM stats computation in the source directly connected to a sink while a more complex configuration with routing, tag manipulation, etc. and get the APM computed at the most efficient place.

The APM stats computation can be seen as a generic way to compute some statistics on a trace flow, thus the following
way forward is suggested:

- Implement a similar logic that the one done in the Datadog OTLP exporter, this would allow user to use multiple

>   Datadog product with Opentelemetry traces and get the same consistent behaviour in all circumstances. APM stats

Reviewer 3: ```suggestion
  Datadog products with Opentelemetry traces and get the same consistent behaviour in all circumstances. APM stats
```

  computation is hooked [there][apm-stats-computation] in the Datadog exporter. But as this is go code it relies on the
  [Agent codebase][agent-code-for-otlp-exporter] to do the [actual computation][agent-handle-span].
- Following the named outputs logic, an `apm_stats` output could be envisionned, this would make the APM stats flow
  explicit and allow another metrics sinks to just tap APM stats as APM stats would be plain Vector metrics.

> - The previous point also implies that the APM stats computation would happend in the source, the code should be fairly

Reviewer 3: ```suggestion
- The previous point also implies that the APM stats computation would happen in the source, the code should be fairly
```

- Support `opentelemetry` source to `datadog_traces` sink forwarding by dealing with:
  - Traces normalization to a single format inside Vector
  - Conversion to/from this format in all traces sources/sinks
- APM stats computation logic, with an implementation for the `opentelemetry` sources, applicable for all traces
  sources. It will come with a knob to turn it on/off.

### Out of scope

N/A

## Pain

- Avoid complex setup when ingesting traces, ultimately pointing every tracing lib directly to Vector should just work
  out-of-the-box with minimal config.

## Proposal

### User Experience

- User would point OpenTelemetry tracing lib directly to a local Vector deployement
- Vector would be configured with a minimal config looking like:

```yaml
sources:
  otlp:
    type: opentelemetry
    address: "[::]:8081"
    mode: grpc
    traces_stats: true # Traces stats would be compliant with APM stats

sinks:
  dd_trace:
    type: datadog_traces
    default_api_key: 12345678abcdef
    inputs:
     - otlp.traces # Would exclusively emit traces
     - otlp.traces_stats # Would exclusively emit metrics

> ```

Reviewer 5: This is clever, and I understand why metrics are generated at the source. My only concern is the footgun around not connecting the `traces_stats` input to the `datadog_traces` sink. It's not obvious or required and I'd expect most users to not do this. Given that the Datadog APM experience is diminished without the stats, we should think about how we can make this automatic or required for the `datadog_traces` sink.

Reviewer 4: Yeah, doing it at the sink level is another option. However, for `datadog_agent` -> `datadog_traces` we do need to propagate the APM stats from the source as the Datadog Agent may not forward all traces and so calculating them in Vector will be inaccurate.

It would be an option for traces coming from the `opentelemetry` source though.

I'm still partial to introducing a transform to generate these stats as a first step. This is flexible enough to cover a myriad of use-cases and still leaves the door open for us to streamline it.

- Support `opentelemetry` source to `datadog_traces` sink forwarding by dealing with:
  - Traces normalization to a single format inside Vector
  - Conversion to/from this format in all traces sources/sinks

> - APM stats computation logic, with an implementation fully function for traces from both the `opentelemetry` and the

Reviewer 3: ```suggestion
- APM stats computation logic, with an implementation fully functional for traces from both the `opentelemetry` and the
```
?
